### ISSUE TITLE
Fix crash: parallel expert training + SetBaggingData + certain Optuna params

### ISSUE THREAD

**Issue by kyo219** [kyo219/LightGBM-MoE#16]
## Problem

When running `benchmark_gate.py` with Optuna, certain parameter combinations cause a crash:

```
[LightGBM] [Warning] Detected that num_threads changed during training (from 4 to 1)
[LightGBM] [Fatal] Check failed: (best_split_info.left_count) > (0)
  at src/treelearner/serial_tree_learner.cpp, line 859
```

## Root Cause (suspected)

Interaction between three features added in this session:

1. **Parallel expert training (#9)**: `OMP_SET_NUM_THREADS(threads_per_expert)` before `std::thread` launch, then restore after join
2. **Sparse activation (#10)**: `SetBaggingData()` restricts histogram construction to per-expert sample subsets
3. **MStepGateLeafReuse (#12)**: Creates per-thread `BinIterator`s and runs OMP parallel leaf traversal

The `num_threads changed during training` warning suggests the thread count set for expert parallel training (Phase 2 of MStepExperts) leaks into MStepGateLeafReuse, causing the tree learner to see inconsistent thread counts. When an expert gets very few samples (Optuna tries extreme params like `min_data_in_leaf=100` with small data), the split check `left_count > 0` fails.

## Reproduction

```bash
python examples/benchmark_gate.py --trials 10 --rounds 50
```

Crashes during MoE leaf_reuse Optuna search on Hamilton dataset (500 samples).

## What works

- `mixture_gate_type=leaf_reuse` with `num_threads=1` — OK up to 100 rounds
- `mixture_gate_type=leaf_reuse` with `num_threads=4`, no Optuna — OK (fixed params)
- `mixture_gate_type=gbdt` and `mixture_gate_type=none` — OK with Optuna

## Likely fix

1. Ensure `OMP_SET_NUM_THREADS(total_threads)` is fully restored before MStepGateLeafReuse runs (currently restored at end of MStepExperts, but the gate retrain interval path inside leaf_reuse calls MStepGate which also does expert-like training)
2. Add a guard in sparse activation: if expert sample count < `min_data_in_leaf`, skip SetBaggingData for that expert (use full dataset with zero gradients instead)
3. Consider removing the trivial zero-gradient `TrainOneIter` call in MStepGateLeafReuse's non-retrain path — it may conflict with the thread count

## Files

- `src/boosting/mixture_gbdt.cpp` (MStepExperts Phase 2, MStepGateLeafReuse)
- `src/treelearner/serial_tree_learner.cpp:859` (crash site)

**Comment by kyo219** [comment]
Fixed in fdba28fb (README) and 29aaa180 (code).

**Root cause**: Optuna picking `bagging_fraction<1.0` / `bagging_freq>0` caused `GBDT::TrainOneIter`'s internal `Bagging()` to overwrite the per-expert `SetBaggingData` from `MStepExperts`. Combined with sparse gradients (zero for unassigned samples in hard M-step), this produced degenerate per-bin histograms whose "best" split had one side empty — tripping `CHECK_GT(left/right_count, 0)` in `serial_tree_learner.cpp:859/869`. Standard GBDT was unaffected because every sample carries a real gradient there.

**Fix layers**:
1. Disable per-expert bagging — sparse activation is the MoE equivalent of bagging, double-bagging is harmful
2. Persist sparse-activation buffers as class members (the previous stack-allocated `expert_sample_indices` left a dangling pointer in `data_partition_->used_data_indices_`)
3. Drop `std::thread` expert parallelism + `OMP_SET_NUM_THREADS` global mutate — concurrent OMP regions from non-OMP host threads were segfaulting in libgomp
4. Wrap each expert's `TrainOneIter` in try/catch as a defensive boundary
5. `MStepGateLeafReuse` perf: flat `vector<BinIterator*>` indexed by inner_feat (was per-thread `unordered_map`), member-cached scratch buffers
6. Drop trivial zero-gradient `gate_->TrainOneIter` on non-retrain iterations of `MStepGateLeafReuse`

**Verification (300 trials × 100 rounds, n_jobs=6 on 24-core)**:
- 0 `CHECK_GT` failures
- 0 segfaults
- 0 "num_threads changed" warnings
- MoE_gbdt RMSE 3.87 vs Standard 5.13 on Synthetic (+24.5%)

README updated with gate-type benchmark and recommended Optuna parallelism setup.
